![1001 genomes](http://1001genomes.org/layout_files/logo_1001genomes_v5.png)  

# *Arabidopsis thaliana* 1001 Genomes Project

[TOC levels=1-3]: # ""

- [*Arabidopsis thaliana* 1001 Genomes Project](#arabidopsis-thaliana-1001-genomes-project)
  - [基本信息](#基本信息)
  - [项目信息](#项目信息)
  - [数据下载](#数据下载)
  - [采用的倍数因子值](#采用的倍数因子值)
  - [Symlink](#symlink)
  - [Run](#run)
  - [Pack and clean](#pack-and-clean)


## 基本信息

* Genome: GCF_000001735.3, TAIR10, 119.668 Mb
* Chloroplast: [NC_000932](https://www.ncbi.nlm.nih.gov/nuccore/NC_000932), **Columbia**, 154478 bp
* Chloroplast: [KX551970](https://www.ncbi.nlm.nih.gov/nuccore/KX551970), **Landsberg erecta**,
  154515 bp
* Mitochondrion: [Y08501](https://www.ncbi.nlm.nih.gov/nuccore/Y08501), 366924 bp


## 项目信息 

* PRJNA273563

> The 1001 Genomes Project was launched at the beginning of 2008 to discover the whole-genome
> sequence variation in 1001 accessions of the reference plant Arabidopsis thaliana. The resulting
> information is paving the way for a new era of genetics that identifies alleles underpinning
> phenotypic diversity across the entire genome and the entire species. Each of the accessions in
> the 1001 Genomes project is an inbred line with seeds that are freely available from the stock
> centre to all our colleagues. Unlimited numbers of plants with identical genotype can be grown and
> phenotyped for each accession, in as many environments as desired, and so the sequence information
> we collect can be used directly in association studies at biochemical, metabolic, physiological,
> morphological, and whole plant-fitness levels. The analyses enabled by this project will have
> broad implications for areas as diverse as evolutionary sciences, plant breeding and human
> genetics.

> 1001 基因组计划于 2008 年初启动, 旨在发现参考植物拟南芥中1001个基因组序列的变异.
> 由此产生的信息正在为一个新的遗传学时代铺平道路, 该时代将识别支撑整个基因组和整个物种表型多样性的等位基因.
> 1001 基因组计划的每一项研究都是一种自交系, 种子可以从库存中心免费获得给我们所有的同事.
> 无限数量的具有相同基因型的植物可以在任意多的环境中生长和表型,
> 因此我们收集的序列信息可以直接用于生物化学、代谢、生理、形态学和整个植物适应性水平的关联研究.
> 该项目的分析将对进化科学、植物育种和人类遗传学等不同领域产生广泛的影响.

## 数据下载

* Reference
  
```shell script
mkdir -p ~/data/plastid/1001/genome
cd ~/data/plastid/1001/genome

for ACCESSION in "NC_000932" "Y08501"; do
    URL=$(printf "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nucleotide&rettype=%s&id=%s&retmode=text" "fasta" "${ACCESSION}");
    curl $URL -o ${ACCESSION}.fa
done

TAB=$'\t'
cat <<EOF > replace.tsv
NC_000932${TAB}Pt
Y08501${TAB}Mt
EOF

cat NC_000932.fa Y08501.fa |
    faops filter -s stdin stdout |
    faops replace stdin replace.tsv stdout |
    faops order stdin <(echo Pt; echo Mt) genome.fa

# bowtie2 index
bowtie2-build --threads 20 genome.fa genome.fa

# chr.sizes
faops size genome.fa > chr.sizes

```

* Illumina

[*Arabidopsis thaliana* 1001 Genomes Project](../ena.md#arabidopsis-thaliana-1001-genomes-project)

* Rsync to hpcc

```shell script
rsync -avP \
    ~/data/plastid/1001/ \
    wangq@202.119.37.251:data/plastid/1001

# rsync -avP wangq@202.119.37.251:data/plastid/1001/ ~/data/plastid/1001

```

## 采用的倍数因子值

* `2`

## Symlink

```shell script
cd ~/data/plastid/1001/

export FOLD=2
export GENOME_SIZE=$(
    cat ~/data/plastid/genome/col_0/chr.sizes |
        tsv-summarize --sum 2
)

cat ena/ena_info.csv |
    mlr --icsv --otsv cat |
    tsv-select -H -f name,srr,base |
    perl -nla -F'\t' -e '
        /^name/ and next;
        my $bases = $F[2];
        $bases =~ s/G$//;
        my $cutoff = $bases * 1000 * 1000 * 1000 / $ENV{GENOME_SIZE} * $ENV{FOLD};
        $cutoff = int $cutoff;
        print join qq(\t), ($F[0], $F[1], $cutoff);
    ' \
    > opts.tsv

cat opts.tsv |
    parallel --colsep '\t' --no-run-if-empty --linebuffer -k -j 1 '
        mkdir -p {1}/1_genome
        pushd {1}/1_genome
    
        ln -fs ../../genome/genome.fa genome.fa
        popd
        
        mkdir -p {1}/2_illumina
        pushd {1}/2_illumina
        
        ln -fs ../../ena/{2}_1.fastq.gz R1.fq.gz
        ln -fs ../../ena/{2}_2.fastq.gz R2.fq.gz
        popd
    '

```

## Run

```shell script
cd ~/data/plastid/1001/

cat opts.tsv | head -n 50 | tail -n 10 |
    parallel --colsep '\t' --no-run-if-empty --linebuffer -k -j 1 '
        cd {1}
        
        rm *.sh
        anchr template \
            --genome 1000000 \
            --parallel 24 \
            --xmx 80g \
            \
            --fastqc \
            --insertsize \
            --kat \
            \
            --trim "--dedupe --cutoff {3} --cutk 31" \
            --qual "25" \
            --len "60" \
            --filter "adapter artifact" \
            \
            --quorum \
            --merge \
            --ecphase "1 2 3" \
            \
            --cov "40 80 120 160 240 320" \
            --unitigger "superreads bcalm tadpole" \
            --splitp 100 \
            --statp 1 \
            --readl 100 \
            --uscale 50 \
            --lscale 5 \
            --redo \
            \
            --extend
            
        bsub -q mpi -n 24 -J "{1}" "
            bash 0_master.sh
        "
    '

```

## Pack and clean

```shell script
cd ~/data/plastid/1001/

cat opts.tsv | head -n 10 |
    parallel --colsep '\t' --no-run-if-empty --linebuffer -k -j 4 '
        cd {1}
        
        if [ -d 4_down_sampling ]; then
            rm -fr 4_down_sampling
            rm -fr 6_down_sampling
        fi
    '

cat opts.tsv | head -n 30 |
    parallel --colsep '\t' --no-run-if-empty --linebuffer -k -j 2 '        
        if [ -f {1}.tar.gz ]; then
            exit;
        fi

        if [ ! -f {1}/7_merge_anchors/anchor.merge.fasta ]; then
            exit;
        fi
        
        echo "==> Create {1}.tar.gz"

        tar -czvf {1}.tar.gz \
            {1}/2_illumina/fastqc \
            {1}/2_illumina/insert_size \
            {1}/2_illumina/kat \
            {1}/2_illumina/trim/Q25L60/pe.cor.fa.gz \
            {1}/2_illumina/trim/Q25L60/env.json \
            {1}/7_merge_anchors/anchor.merge.fasta \
            {1}/7_merge_anchors/others.non-contained.fasta \
            {1}/8_megahit/anchor/anchor.fasta \
            {1}/8_megahit/megahit.non-contained.fasta \
            {1}/8_mr_megahit/anchor/anchor.fasta \
            {1}/8_mr_megahit/megahit.non-contained.fasta \
            {1}/8_spades/anchor/anchor.fasta \
            {1}/8_spades/spades.non-contained.fasta \
            {1}/8_mr_spades/anchor/anchor.fasta \
            {1}/8_mr_spades/spades.non-contained.fasta \
            {1}/8_platanus/anchor/anchor.fasta \
            {1}/8_platanus/platanus.non-contained.fasta \
            {1}/9_quast \
            {1}/*.md
    '

cat opts.tsv | head -n 30 |
    parallel --colsep '\t' --no-run-if-empty --linebuffer -k -j 4 '
        rm -fr {1}
    '

```
